### DISCUSSION TITLE
Setting branch and revision on a url in a resource

### DISCUSSION THREAD

**Discussion by nwhetsell** [Homebrew/discussions#2595] [Writing Formulae/Casks]
Currently, the `csound` formula cannot be built from source:

```
% brew install --build-from-source --verbose csound
==> Downloading https://github.com/Ableton/link/archive/Link-3.0.3.tar.gz
Already downloaded: /Users/nathanwhetsell/Library/Caches/Homebrew/downloads/004f5de8cddf15ddf13626a63321a6a4d8b4622468c38b6219a4063a64412862--link-Link-3.0.3.tar.gz
==> Verifying checksum for '004f5de8cddf15ddf13626a63321a6a4d8b4622468c38b6219a4063a64412862--link-Link-3.0.3.tar.gz'
==> Cloning https://github.com/csound/plugins.git
Updating /Users/nathanwhetsell/Library/Caches/Homebrew/csound--csound-plugins--git
git config remote.origin.url https://github.com/csound/plugins.git
git remote set-head origin --auto
error: Not a valid ref: refs/remotes/origin/develop
origin/HEAD set to develop
Error: csound: Failed to download resource "csound--csound-plugins"
Failure while executing; `git remote set-head origin --auto` exited with 1. Here's the output:
error: Not a valid ref: refs/remotes/origin/develop
origin/HEAD set to develop
```

This is because the `csound` formula uses another Git repository at a particular commit as a resource, but this commit is no longer on the default branch (“develop”, in this case):

https://github.com/Homebrew/homebrew-core/blob/7734cc10c83cfff1621037fd71e127b7bbe2fb1c/Formula/csound.rb#L59-L62

Changing the resource block to—

```ruby
resource "csound-plugins" do
    url "https://github.com/csound/plugins.git",
        branch:   "master",
        revision: "63b784625e66109babd3b669abcb55f5b404f976"
  end
```

—enables the formula to build, but incorrectly. What seems to be happening is that Homebrew checks out the specified branch, but doesn’t move to the specified revision:

```
==> Cloning https://github.com/csound/plugins.git
Updating /Users/nathanwhetsell/Library/Caches/Homebrew/csound--csound-plugins--git
git config remote.origin.url https://github.com/csound/plugins.git
git config remote.origin.fetch \+refs/heads/master:refs/remotes/origin/master
git config remote.origin.tagOpt --no-tags
git config advice.detachedHead false
git fetch origin
==> Checking out branch master
git checkout -f master --
Already on 'master'
Your branch is up to date with 'origin/master'.
git reset --hard origin/master --
HEAD is now at b32f5e1 released code
```

That last line is unexpected; I’d expect HEAD to be set, at some point, to `63b7846` (the revision in the resource block).

Is there a way to specify both a branch and a revision for a Git repository in a resource block?

**Comment by ipatch** [comment]
```
brew install [FORMULA] -s
```

is different from

```
brew install [FORMULA] --HEAD
```

think of it this way,

you could download a tarball of pkg/formula ie. llvm for example that will be set to a specific revision, but you'd still need to build that tarball from source ie. the `-s` flag. but you can also build the tip of the llvm project ie. `HEAD`

so `HEAD` and `-s` ie. source are not exactly the same thing.

**Reply by nwhetsell** [reply]
@ipatch Thanks, but I’m afraid you misunderstood what I’m asking about. I (thoroughly) understand the difference between `-s`/`--build-from-source` and `--HEAD`. I’m asking about an issue that’s occurring when installing Csound with `--build-from-source`, not `--HEAD`. (Actually, this issue is probably also occurring with `--HEAD`, but I haven’t checked.)

The Csound formula uses a Git repository in a resource block (the resource Git repository – https://github.com/csound/plugins.git – is **not** the same as the repository for Csound itself). Until recently, specifying a revision (a commit hash) for the resource Git repository worked; Homebrew checked out the repo at the specified commit. At some point, the default branch of the resource Git repository changed, and specifying the commit stopped working because that commit is on a different branch. Currently, I seem to be able to set the branch of the resource Git repository, but not the commit.

Without the correct commit for the resource Git repository, Csound builds, but incorrectly. This is somewhat problematic because seemingly minor changes can’t be made to the Csound formula because its test fails, as occurred [here](https://github.com/Homebrew/homebrew-core/runs/4423583294?check_suite_focus=true).

**Reply by ipatch** [reply]
okay, (correct me if i'm wrong on this please). you want the csound plugins to be included within the specific formula file / formulae you linked above. ie. `csound` that formulae uses the csound plugins a resource block to build or download the plugins as a resource for csound.

1. taking a quick look at the csound plugins it looks like they can be builtin into their own project using cmake. looks the plugins build and install using the usual cmake dance.

that said if want those plugins that specific commit / revision you link above ie. `63b784625e66109babd3b669abcb55f5b404f976`

why not change the URL to point to that specific commit in the repos history? and then just grab the plugins at that point in the repo history.

```
https://github.com/csound/plugins/archive/63b784625e66109babd3b669abcb55f5b404f976.zip
```

would that work?

---

i did a quick search for `.zip` in the homebrew-core repo and saw the `azure-cli.rb` formula using several resource blocks with `.zip` extensions and providing a sha256sum of the zip file.

let me know if that would work.

**Reply by nwhetsell** [reply]
That would work, thank you!

**Reply by SMillerDev** [reply]
While a branch defines a list of commits, as far as I know nothing requires you to search for a commit in a specific branch. So it's likely your problem isn't with branches at all.
